File: apps/tabtin-ios/Tabtin/Core/Workspace/CloudDocsShare.swift
import Foundation

/// 云文档公开分享支持的资源类型。
///
/// 刻意只覆盖 tabdoc / tabdata。tabfiles 没有公开链接，不要套同一套 URL。
enum CloudShareResourceType: String, Sendable, Hashable {
    case document
    case table

    /// 公网范围的 share_type 取值：doc 是 `"public"`，table 是 `"data"`。
    /// 这是后端的既有不对称——写死一边会让另一种资源静默开成错误范围。
    var anyoneWireValue: String {
        switch self {
        case .document: return "public"
        case .table: return "data"
        }
    }

    /// 公开分享页路径段：`/shared/docs|tables/{shareId}`。
    var publicPathSegment: String {
        switch self {
        case .document: return "docs"
        case .table: return "tables"
        }
    }

    /// 后端支持、且 UI 应展示的权限档位。table 没有 comment。
    var availablePermissions: [CloudSharePermission] {
        switch self {
        case .document: return [.view, .comment, .edit]
        case .table: return [.view, .edit]
        }
    }

    /// 从归一后的类型名解析。只认 `tabdoc` / `tabdata`。
    static func from(normalizedType: String) -> CloudShareResourceType? {
        switch normalizedType {
        case "tabdoc": return .document
        case "tabdata": return .table
        default: return nil
        }
    }
}

enum CloudShareScope: Sendable, Hashable, CaseIterable {
    case organization
    case anyone

    func wireValue(for type: CloudShareResourceType) -> String {
        switch self {
        case .organization: return "organization"
        case .anyone: return type.anyoneWireValue
        }
    }

    /// 反解 share_type。认不出来时返回 organization（保守，宁可显示成范围更小）。
    static func from(wireValue: String, type: CloudShareResourceType) -> CloudShareScope {
        if wireValue == "organization" { return .organization }
        if wireValue == type.anyoneWireValue { return .anyone }
        return .organization
    }
}

enum CloudSharePermission: String, Sendable, Hashable {
    case view
    case comment
    case edit
}

/// 后端 ShareOut / DataShareOut 的交集字段。
///
/// `allow_download` / `allow_copy` 本期不给 UI，先不解码。
/// TabData 的 DataShareOut 可能缺 `organization_id` / `is_active`，缺省要能解。
struct CloudDocShare: Decodable, Sendable, Hashable {
    let shareId: String
    let shareType: String
    let permission: String
    let hasPassword: Bool
    let expireAt: String?
    let organizationId: String?
    let visitCount: Int?
    let isActive: Bool
    let createdAt: String?

    enum CodingKeys: String, CodingKey {
        case permission
        case shareId = "share_id"
        case shareType = "share_type"
        case hasPassword = "has_password"
        case expireAt = "expire_at"
        case organizationId = "organization_id"
        case visitCount = "visit_count"
        case isActive = "is_active"
        case createdAt = "created_at"
    }

    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        shareId = try container.decode(String.self, forKey: .shareId)
        shareType = try container.decode(String.self, forKey: .shareType)
        permission = try container.decode(String.self, forKey: .permission)
        hasPassword = try container.decodeIfPresent(Bool.self, forKey: .hasPassword) ?? false
        expireAt = try container.decodeIfPresent(String.self, forKey: .expireAt)
        organizationId = try container.decodeIfPresent(String.self, forKey: .organizationId)
        visitCount = try container.decodeIfPresent(Int.self, forKey: .visitCount)
        // TabData DataShareOut 无 is_active；能读到分享对象即视为生效中。
        isActive = try container.decodeIfPresent(Bool.self, forKey: .isActive) ?? true
        createdAt = try container.decodeIfPresent(String.self, forKey: .createdAt)
    }
}

enum CloudDocsShareError: Error, Sendable, Equatable {
    case forbidden
    case publicExposureNotAcknowledged
    case unsupportedResourceType
    case other(String)
}

// MARK: - 信封（data 内层，APIClient 已剥掉外层 success）

struct CloudDocShareFetchResponse: Decodable, Sendable {
    let share: CloudDocShare?
    let enabled: Bool?
}

struct CloudDocShareMutationResponse: Decodable, Sendable {
    let share: CloudDocShare
}

struct CloudDocShareDisableResponse: Decodable, Sendable {
    let disabledCount: Int?

    enum CodingKeys: String, CodingKey {
        case disabledCount = "disabled_count"
    }
}

struct CloudDocsCollaborator: Decodable, Identifiable, Sendable, Hashable {
    let userId: String
    let nickname: String
    let avatar: String?
    let email: String
    let permission: String
    var id: String { userId }

    /// 协作者接口使用 `editor` / `admin`；兼容早期环境曾返回的 `edit`，
    /// 但不把只读成员算作维护者。
    var canEdit: Bool {
        switch permission.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "admin", "editor", "edit": return true
        default: return false
        }
    }

    enum CodingKeys: String, CodingKey { case userId = "user_id", nickname, avatar, email, permission }
}

/// 协作者接口把 owner 单独放在顶层，且 owner 没有 permission 字段。
struct CloudDocsResourceOwner: Decodable, Identifiable, Sendable, Hashable {
    let userId: String
    let nickname: String
    let avatar: String?
    let email: String
    var id: String { userId }
    enum CodingKeys: String, CodingKey { case userId = "user_id", nickname, avatar, email }
}

struct CloudDocsCollaboratorList: Decodable, Sendable {
    let owner: CloudDocsResourceOwner?
    let collaborators: [CloudDocsCollaborator]
}

struct CloudDocsMutationAck: Decodable, Sendable {}

File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/data/model/CloudDocsShare.kt
package com.tabtin.mobile.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * 云文档公开分享支持的资源类型。
 *
 * 刻意只覆盖 tabdoc / tabdata。tabfiles 没有公开链接，不要套同一套 URL。
 */
public enum class CloudShareResourceType {
    DOCUMENT,
    TABLE,
    ;

    /** 公网范围的 share_type：doc 是 `"public"`，table 是 `"data"`。 */
    public val anyoneWireValue: String
        get() = when (this) {
            DOCUMENT -> "public"
            TABLE -> "data"
        }

    /** 公开分享页路径段：`/shared/docs|tables/{shareId}`。 */
    public val publicPathSegment: String
        get() = when (this) {
            DOCUMENT -> "docs"
            TABLE -> "tables"
        }

    /** 后端支持、且 UI 应展示的权限档位。table 没有 comment。 */
    public val availablePermissions: List<CloudSharePermission>
        get() = when (this) {
            DOCUMENT -> listOf(
                CloudSharePermission.VIEW,
                CloudSharePermission.COMMENT,
                CloudSharePermission.EDIT,
            )
            TABLE -> listOf(
                CloudSharePermission.VIEW,
                CloudSharePermission.EDIT,
            )
        }

    public companion object {
        /** 从归一后的类型名解析。只认 `tabdoc` / `tabdata`。 */
        public fun fromNormalizedType(normalizedType: String): CloudShareResourceType? =
            when (normalizedType) {
                "tabdoc" -> DOCUMENT
                "tabdata" -> TABLE
                else -> null
            }
    }
}

public enum class CloudShareScope {
    ORGANIZATION,
    ANYONE,
    ;

    public fun wireValue(type: CloudShareResourceType): String = when (this) {
        ORGANIZATION -> "organization"
        ANYONE -> type.anyoneWireValue
    }

    public companion object {
        /**
         * 反解 share_type。认不出来时返回 organization（保守，宁可显示成范围更小）。
         */
        public fun fromWireValue(wireValue: String, type: CloudShareResourceType): CloudShareScope =
            when (wireValue) {
                "organization" -> ORGANIZATION
                type.anyoneWireValue -> ANYONE
                else -> ORGANIZATION
            }
    }
}

public enum class CloudSharePermission(public val wireValue: String) {
    VIEW("view"),
    COMMENT("comment"),
    EDIT("edit"),
    ;

    public companion object {
        public fun fromWireValue(wireValue: String): CloudSharePermission? =
            entries.firstOrNull { it.wireValue == wireValue }
    }
}

/**
 * 后端 ShareOut / DataShareOut 的交集字段。
 *
 * `allow_download` / `allow_copy` 本期不给 UI，先不解码。
 * TabData 的 DataShareOut 可能缺 `organization_id` / `is_active`，缺省要能解。
 */
@Serializable
public data class CloudDocShare(
    @SerialName("share_id") val shareId: String,
    @SerialName("share_type") val shareType: String,
    val permission: String,
    @SerialName("has_password") val hasPassword: Boolean = false,
    @SerialName("expire_at") val expireAt: String? = null,
    @SerialName("organization_id") val organizationId: String? = null,
    @SerialName("visit_count") val visitCount: Int? = null,
    /** TabData DataShareOut 无 is_active；能读到分享对象即视为生效中。 */
    @SerialName("is_active") val isActive: Boolean = true,
    @SerialName("created_at") val createdAt: String? = null,
)

public sealed class CloudDocsShareError(message: String? = null) : Exception(message) {
    public data object Forbidden : CloudDocsShareError("forbidden")
    public data object PublicExposureNotAcknowledged :
        CloudDocsShareError("public_exposure_not_acknowledged")
    public data object UnsupportedResourceType : CloudDocsShareError("unsupported_resource_type")
    public data class Other(val detail: String) : CloudDocsShareError(detail)
}

@Serializable
public data class CloudDocShareFetchResponse(
    val share: CloudDocShare? = null,
    val enabled: Boolean? = null,
)

@Serializable
public data class CloudDocShareMutationResponse(
    val share: CloudDocShare,
)

@Serializable
public data class CloudDocShareDisableResponse(
    @SerialName("disabled_count") val disabledCount: Int? = null,
)

/** 资源级协作者。权限值沿用后端 `viewer` / `editor` / `admin` 契约。 */
@Serializable
public data class CloudDocsCollaborator(
    @SerialName("user_id") val userId: String,
    val nickname: String = "",
    val avatar: String? = null,
    val email: String = "",
    val permission: String = "viewer",
)

@Serializable
public data class CloudDocsOwner(
    @SerialName("user_id") val userId: String,
    val nickname: String = "",
    val avatar: String? = null,
    val email: String = "",
)

@Serializable
public data class CloudDocsCollaboratorsResponse(
    val owner: CloudDocsOwner,
    val collaborators: List<CloudDocsCollaborator> = emptyList(),
)

@Serializable
public data class CloudDocsInviteRequest(
    @SerialName("user_ids") val userIds: List<String>,
    val permission: String,
)

@Serializable
public data class CloudDocsUpdateCollaboratorRequest(val permission: String)

/** 与 iOS Endpoints.TabDoc / TabData share 路径对齐，便于单测钉住。 */
public object CloudDocsSharePaths {
    public fun documentShare(documentId: String): String =
        "tabdoc/documents/$documentId/share"

    public fun documentShareRefresh(documentId: String): String =
        "tabdoc/documents/$documentId/share/refresh"

    public fun tableShare(tableId: String): String =
        "tabdata/tables/$tableId/share"
}
